src/library/scala/collection/immutable/OrderedMap.scala

```diff
@@ -0,0 +1,518 @@
+    if (orderBy == this.orderedBy) this
+    else new OrderedMap(ordering, mapping, ordinal, orderBy)
+  }
+
+  def updated[V1 >: V](key: K, value: V1): OrderedMap[K, V1] = {
+    mapping.get(key) match {
+      case e if ordinal == -1 && (orderedBy == OrderBy.Modification || e.isEmpty) ⇒
+        // Reinsert into fresh instance to restart ordinal counting, expensive but only done after 2^32 updates.
+        OrderedMap.empty[K, V](orderedBy) ++ this + (key → value)
+      case Some((o, _)) if orderedBy == OrderBy.Insertion =>
+        new OrderedMap(
+          ordering.incl(o, key),
+          mapping.updated[(Int, V1)](key, (o, value)),
+          ordinal, // Do not increment the ordinal since the key was already present, i.e. o <= ordinal.
+          orderedBy)
+      case Some((o, _)) =>
+        val o1 = if (ordinal == Int.MaxValue) Int.MinValue else ordinal + 1
+        new OrderedMap(
+          ordering.excl(o).incl(o1, key),
+          mapping.updated[(Int, V1)](key, (o1, value)),
+          o1,
+          orderedBy)
+      case None ⇒
+        val o1 = ordinal + 1
+        new OrderedMap(
+          ordering.incl(o1, key),
+          mapping.updated[(Int, V1)](key, (o1, value)),
+          o1,
+          orderedBy)
+    }
+  }
+
+  def removed(key: K): OrderedMap[K, V] = {
+    mapping.get(key) match {
+      case Some((o, _)) ⇒
+        new OrderedMap(
+          ordering.excl(o),
+          mapping.removed(key),
+          ordinal,
+          orderedBy)
+      case None ⇒
+        this
+    }
+  }
+
+  def refresh(key: K): OrderedMap[K, V] = {
+    mapping.get(key) match {
+      case Some((o, _)) ⇒
+        val o1 = ordinal + 1
+        new OrderedMap(
+          ordering.excl(o).incl(o1, key),
+          mapping,
+          o1,
+          orderedBy)
+      case None ⇒
+        this
+    }
+  }
+
+  override def contains(key: K): Boolean = mapping.contains(key)
+
+  override def head: (K, V) = binding(ordering.head)
+
+  override def headOption = ordering.headOption.map(binding)
+
+  override def last: (K, V) = binding(ordering.last)
+
+  override def lastOption: Option[(K, V)] = ordering.lastOption.map(binding)
+
+  override def tail: OrderedMap[K, V] = {
+    val (head, tail) = ordering.headTail
+    new OrderedMap(tail, mapping.removed(head), ordinal, orderedBy)
+  }
+
+  override def init: OrderedMap[K, V] = {
+    val (init, last) = ordering.initLast
+    new OrderedMap(init, mapping.removed(last), ordinal, orderedBy)
+  }
+
+  override def slice(from: Int, until: Int): OrderedMap[K, V] = {
+    val sz = size
+    if (sz == 0 || from >= until) empty
+    else {
+      val sz = size
+      val f = if (from >= 0) from else 0
+      val u = if (until <= sz) until else sz
+      if (f >= u) empty
+      else {
+        val (front, rest) = this.ordering.splitAt(f)
+        val (ordering, rear) = rest.splitAt(u - f)
+        var mapping = this.mapping
+        val frontIter = front.iterator
```

Comment:
I have added support for this (if the slice will contain more than half the original entries then the above peel-off-front-and-rear approach is used, otherwise a builder-based approach is used instead). More benchmarking should probably be done to get a more exact pivot-point.